src/sqlancer/clickhouse/ClickHouseSchema.java

```diff
@@ -0,0 +1,313 @@
+        }
+
+        public List<ClickHouseSchema.ClickHouseColumn> getColumns() {
+            return columns;
+        }
+
+        public String columnNamesAsString() {
+            return getColumns().stream().map(t -> t.getTable().getName() + "." + t.getName())
+                    .collect(Collectors.joining(", "));
+        }
+
+        public String columnNamesAsString(Function<ClickHouseSchema.ClickHouseColumn, String> function) {
+            return getColumns().stream().map(function).collect(Collectors.joining(", "));
+        }
+
+        public ClickHouseRowValue getRandomRowValue(Connection con, StateToReproduce.ClickHouseStateToReproduce state) throws SQLException {
+            String randomRow = String.format("SELECT %s, %s FROM %s ORDER BY RANDOM() LIMIT 1", columnNamesAsString(
+                    c -> c.getTable().getName() + "." + c.getName() + " AS " + c.getTable().getName() + c.getName()),
+                    columnNamesAsString(c -> "typeof(" + c.getTable().getName() + "." + c.getName() + ")"),
+                    tableNamesAsString());
+            Map<ClickHouseColumn, ClickHouseConstant> values = new HashMap<>();
+            try (Statement s = con.createStatement()) {
+                ResultSet randomRowValues;
+                try {
+                    randomRowValues = s.executeQuery(randomRow);
+                } catch (SQLException e) {
+                    throw new IgnoreMeException();
+                }
+                if (!randomRowValues.next()) {
+                    throw new AssertionError("could not find random row! " + randomRow + "\n" + state);
+                }
+                for (int i = 0; i < getColumns().size(); i++) {
+                    ClickHouseSchema.ClickHouseColumn column = getColumns().get(i);
+                    int columnIndex = randomRowValues.findColumn(column.getTable().getName() + column.getName());
+                    assert columnIndex == i + 1;
+                    String typeString = randomRowValues.getString(columnIndex + getColumns().size());
+                    ClickHouseDataType valueType = getColumnType(typeString).getType();
+                    ClickHouseConstant constant = getConstant(randomRowValues, columnIndex, valueType);
+                    values.put(column, constant);
+                }
+                assert !randomRowValues.next();
+                state.randomRowValues = values;
+                return new ClickHouseSchema.ClickHouseRowValue(this, values);
+            }
+
+        }
+
+    }
+
+
+
+    public ClickHouseSchema(List<ClickHouseTable> databaseTables) {
+        super(databaseTables);
+    }
+
+    public ClickHouseTables getRandomTableNonEmptyTables() {
+        return new ClickHouseTables(Randomly.nonEmptySubset(getDatabaseTables()));
+    }
+
+    private static ClickHouseLancerDataType getColumnType(String typeString) {
+        ClickHouseLancerDataType primitiveType;
+//        if (typeString.startsWith("Nullable(")) {
```

Comment:
If possible, avoid checking in commented-out code. We do not enforce this currently, since SQLancer's code contains already many instances, but I plan to remove all commented-out code at a later stage.